dll/win32/advapi32/wine/security.c

```diff
@@ -1696,27 +1696,237 @@ AccessCheck(IN PSECURITY_DESCRIPTOR pSecurityDescriptor,
+ * A pointer to a security descriptor that identifies the security
+ * information of an object being accessed. This function walks
+ * through this descriptor for any ACLs and respective access
+ * rights if access can be granted.
+ *
+ * @param[in] PrincipalSelfSid
+ * A pointer to a principal self SID. This parameter can be NULL if
+ * the associated object being checked for access does not represent
+ * a principal.
+ *
+ * @param[in] ClientToken
+ * A handle to an access token, that identifies the client of which
+ * requests access to the target object.
+ *
+ * @param[in] DesiredAccess
+ * The access right bitmask where the client wants to acquire. This
+ * can be an OR'ed set of multiple access rights or MAXIMUM_ALLOWED
+ * to request all of possible access rights the target object allows.
+ * If only some rights were granted but not all the access is deemed
+ * as denied.
+ *
+ * @param[in] ObjectTypeList
+ * A pointer to a given object type list. If this parameter is not NULL
+ * the function will perform an access check against the main object
+ * and sub-objects of this list. If this parameter is NULL and
+ * ObjectTypeListLength is 0, the function will perform a normal
+ * access check instead.
+ *
+ * @param[in] ObjectTypeListLength
+ * The length of the object type list array, pointed by ObjectTypeList.
+ * This length in question represents the number of elements in such array.
+ * This parameter must be 0 if no array list is provided.
+ *
+ * @param[in] GenericMapping
+ * The generic mapping of access rights of an object type.
+ *
+ * @param[out] PrivilegeSet
+ * A pointer to a set of privileges that were used to perform the
+ * access check, returned to caller. This function will return no
+ * privileges (privilege count set to 0) if no privileges were used
+ * to accomplish the access check. This parameter must not be NULL!
+ *
+ * @param[in,out] PrivilegeSetLength
+ * The total length size of a set of privileges. This length represents
+ * the count of elements in the privilege set array.
+ *
+ * @param[out] GrantedAccess
+ * A pointer to granted access rights, returned to the caller.
+ *
+ * @param[out] AccessStatus
+ * A pointer to a boolean value that indicates whether access is granted
+ * or denied to the client that requests access to the entire hierarchy
+ * of an object type list. If ObjectTypeList is NULL, this value represents
+ * the access that is granted or denied to the target object, just like
+ * in AccessCheck.
+ *
+ * @return
+ * The function returns TRUE if the access check operation has completed
+ * successfully, otherwise it returns FALSE.
+ */
+BOOL
+WINAPI
+AccessCheckByType(
+    _In_ PSECURITY_DESCRIPTOR pSecurityDescriptor,
+    _In_opt_ PSID PrincipalSelfSid,
+    _In_ HANDLE ClientToken,
+    _In_ DWORD DesiredAccess,
+    _In_opt_ POBJECT_TYPE_LIST ObjectTypeList,
+    _In_ DWORD ObjectTypeListLength,
+    _In_ PGENERIC_MAPPING GenericMapping,
+    _Out_ PPRIVILEGE_SET PrivilegeSet,
```

Comment:
> This one is claimed to be _Out_opt_ by MSDN. Is this correct?

No it isn't. MSDN documentations are misleading at best, the caller is expected to provide a privilege set variable of which the functions (`AccessCheck` and `ByType**` variants) will return a set of privileges that a client currently holds in UM. We do have tests that back up this fact. The exception to this rule is `SeAccessCheck` as one could freely pass `PrivilegeSet` as NULL**.

The reasoning behind this is the scope of `PRIVILEGE_SET` -- this structure serves the purpose to hold privilege information about an untrusted entity, in this case the UM. In KM you're the utmost privileged entity therefore collecting `PRIVILEGE_SET` data would be extremely pointless.

I should probably provide a PR to Microsoft Docs to clarify all of this stuff.

** -  While you could call `SeAccessCheck` in both KM and UM, in KM the function will not return any set of privileges due to reasons as stated above. Whereas in UM it will return a set of privileges that are granted to the client whom requests access to an object. The `SeAccessCheck` semantics allow you to optionally pass `PrivilegeSet`, while `NtAccessCheck`, `NtAccessCheckByType` and `NtAccessCheckByTypeResultList` do not.